src/ADCS.SidExtension.PolicyModule/DsUtils.cs

```diff
@@ -57,13 +57,24 @@ static class DsUtils {
 
         if (name.Contains("@")) {
             // upn
-            Int32 splitIndex = name.IndexOf("@", StringComparison.Ordinal);
-            domainTokens = name.Substring(splitIndex + 1).Split('.');
+            String domainPart = name.Substring(splitIndex + 1);
+            using (var forest = Forest.GetCurrentForest()) {
+                using (var domain = forest.Domains.FirstOrDefault(d => d.Forest.Name.Equals(domainPart, StringComparison.OrdinalIgnoreCase))) {
```

Comment:
I'm not sure how it is supposed to work. All domains share same forest name, so `forest.Domain.FirstOrDefault` call makes little to no sense, because first domain entry will match the predicate and others are not evaluated. And there always will be one object -- forest root domain. In multi-domain forest it is going to fail. IIRC, domains in `forest.Domains` collection are ordered by name, so the predicate will match the first alphabetically ordered domain in the list. And it doesn't necessary have to be a domain where this particular UPN lives in.

In other words, I understand the problem, however this PR addresses the issue in a single-domain forest. Once you get two domains, it is going to fail.

In case of custom UPN suffix, it may be necessary to query all domains in a forest until match is found.